docs/internals/data_file.md

```diff
@@ -0,0 +1,198 @@
+---
+sidebar_position: 2
+---
+
+# Data File
+
+> “Just show me the tables already!”
+> — probably not Fred Brooks
+
+Each TigerBeetle replica stores all data inside a single file, called the data file (conventional
+extension is `.tigerbeetle`). This document describes the high level layout of the data file. The
+presentation is simplified a bit, to provide intuition without drowning the reader in details.
+Consult the source code for byte-level details!
+
+The data file is divided into several zones, with the main ones being:
+
+- write-ahead log
+- superblock
+- grid
+
+The grid forms the bulk of the data file (up to several terabytes). It is an elastic array of 64KiB
+blocks:
+
+```zig
+pub const Block = [constants.block_size]u8;
+pub const BlockPtr = *align(constants.sector_size) Block;
+```
+
+The grid serves as a raw storage layer. Higher level data structures (notably, the LSM tree) are
+mapped to physical grid blocks. Because TigerBeetle is deterministic, the used portion of the grid
+is identical across up-to-date replicas. This storage determinism is exploited to implement state
+sync and repair on the level of grid blocks, see [the repair protocol](./vsr.md#protocol-repair-grid).
+
+A grid block is identified by a pair of a `u64` index and `u128` checksum:
+
+```zig
+pub const BlockReference = struct {
+    index: u64,
+    checksum: u128,
+};
+```
+
+The block checksum is stored outside of the block itself, to protect from misdirected writes. So, to
+read a block, you need to know the block's index and checksum from "elsewhere", where "elsewhere" is
+either a different block, or the superblock. Overall, the grid is used to implement a purely
+functional, persistent (in both senses), garbage collected data structure which is updated
+atomically by swapping the pointer to the root node. This is the classic copy-on-write technique
+commonly used in filesystems. In fact, you can think of TigerBeetle's data file as a filesystem.
```

Comment:
Either split "filesystems" into 2 words, or add it to the spell checker. I think both are correct.